### DISCUSSION TITLE
How can I integrate react-number-format into react-hook-form with the MUI text field?

### DISCUSSION THREAD

**Discussion by bpeter340** [react-hook-form/react-hook-form#10138] [Q&A] [answered]
I'm having a validation issue when I try to integrate react-hook-form with react-number-format. I can get the input field state, but validation only works for empty fields despite character length checks. What could be the issue here?

````
<Controller
    name="phone_number"
    control={control}
    render={({ field, fieldState }) => (
      <PatternFormat
        {...field}
        label="Phone Number"
        customInput={TextField}
        type="tel"
        format="###-###-####"
        autoComplete="off"
        fullWidth
        error={!!fieldState.error?.message}
        helperText={fieldState.error?.message}
      />
    )}
  />
````

Here's a CodePen demo that demonstrates the issue:  https://codesandbox.io/s/blazing-hill-9pn3m4?file=/src/App.js

**Comment by Sajarin-M** [comment, marked as answer]
**There are two methods to fix this issue:**

1. If you need to get the formatted value in your form state, the `length` method in Yup doesn't work as expected in this case because, as soon as we start typing in the field, the length becomes 12, including white spaces and hyphens. Therefore, you can use your own validation logic with the `test` method by replacing all white spaces and hyphens in this case, like this:
```js
 phone_number: Yup.string()
  .required("Phone Number is required")
  .test(
    "test-phone-length",
    "Must be a valid phone number (10 digits)",
    (value) => value.replace(/[\s-]+/g, "").length === 10
  )
```
<br />

2. If you don't need to save the formatted value in your form state, you can change your component definition to save the `floatValue` from `onValueChange` event to form state, and your previous validation with `length` should work, like this:
```js
<PatternFormat
  getInputRef={field.ref}
  onBlur={field.onBlur}
  value={field.value}
  label="Phone Number"
  customInput={TextField}
  type="tel"
  onValueChange={(value) => {
    console.log(value);
    field.onChange(value.floatValue);
  }}
  format="###-###-####"
  autoComplete="off"
  fullWidth
  error={!!fieldState.error?.message}
  helperText={fieldState.error?.message}
  />
```

In either case, the `ref` prop is not available on `PatternFormat`. Instead, it uses `getInputRef`, so you need to pass the field ref as `getInputRef={field.ref}`.
